oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/access/BearerTokenAccessDeniedHandler.java

```diff
@@ -0,0 +1,134 @@
+import org.springframework.security.web.access.AccessDeniedHandler;
+import org.springframework.util.StringUtils;
+
+/**
+ * Translates any {@link AccessDeniedException} into an HTTP response in accordance with
+ * <a href="https://tools.ietf.org/html/rfc6750#section-3" target="_blank">RFC 6750 Section 3: The WWW-Authenticate</a>.
+ *
+ * So long as the class can prove that the request has a valid OAuth 2.0 {@link Authentication}, then will return an
+ * <a href="https://tools.ietf.org/html/rfc6750#section-3.1" target="_blank">insufficient scope error</a>; otherwise,
+ * it will simply indicate the scheme (Bearer) and any configured realm.
+ *
+ * @author Josh Cummings
+ * @since 5.1
+ */
+public final class BearerTokenAccessDeniedHandler implements AccessDeniedHandler {
+
+	private static final Collection<String> WELL_KNOWN_SCOPE_ATTRIBUTE_NAMES =
+			Arrays.asList("scope", "scp");
+
+	private String realmName;
+
+	/**
+	 * Collect error details from the provided parameters and format according to
+	 * RFC 6750, specifically {@code error}, {@code error_description}, {@code error_uri}, and {@scope scope}.
+	 *
+	 * @param request that resulted in an <code>AccessDeniedException</code>
+	 * @param response so that the user agent can be advised of the failure
+	 * @param accessDeniedException that caused the invocation
+	 *
+	 */
+	@Override
+	public void handle(
+			HttpServletRequest request, HttpServletResponse response,
+			AccessDeniedException accessDeniedException) {
+
+		Map<String, String> parameters = new LinkedHashMap<>();
+
+		if (this.realmName != null) {
+			parameters.put("realm", this.realmName);
+		}
+
+		if (request.getUserPrincipal() instanceof AbstractOAuth2TokenAuthenticationToken) {
+			AbstractOAuth2TokenAuthenticationToken token =
+					(AbstractOAuth2TokenAuthenticationToken) request.getUserPrincipal();
+
+			String scope = getScope(token);
+
+			parameters.put("error", BearerTokenErrorCodes.INSUFFICIENT_SCOPE);
+			parameters.put("error_description",
+					String.format("The token provided has insufficient scope [%s] for this request", scope));
+			parameters.put("error_uri", "https://tools.ietf.org/html/rfc6750#section-3.1");
+
+			if (StringUtils.hasText(scope)) {
+				parameters.put("scope", scope);
+			}
+		}
+
+		String wwwAuthenticate = computeWWWAuthenticateHeaderValue(parameters);
+
+		response.addHeader(HttpHeaders.WWW_AUTHENTICATE, wwwAuthenticate);
+		response.setStatus(HttpStatus.FORBIDDEN.value());
+	}
+
+	/**
+	 * Set the default realm name to use in the bearer token error response
+	 *
+	 * @param realmName
+	 */
+	public final void setRealmName(String realmName) {
+		this.realmName = realmName;
+	}
+
+	private static String getScope(AbstractOAuth2TokenAuthenticationToken token) {
+
+		Map<String, Object> attributes = token.getTokenAttributes();
+
+		for (String attributeName : WELL_KNOWN_SCOPE_ATTRIBUTE_NAMES) {
+			Object scopes = attributes.get(attributeName);
```

Comment:
`scopes` could be `null` so should do `null` check first

Reply:
I believe this is what `instanceof` does for me, no? e.g. A `NullPointerException` is not possible in the ensuing block of code.

Are you suggesting that this makes it more readable? I tend to see unnecessary `null` checks as extra verbosity, of which Java already has plenty. :)